File: server/src/main/java/io/github/mufasa1976/calcmaster/records/CalculationProperties.java
package io.github.mufasa1976.calcmaster.records;

import io.github.mufasa1976.calcmaster.enums.Operator;
import lombok.Builder;

import java.util.List;
import java.util.Optional;

@Builder
public record CalculationProperties(
    List<Operator> operators,
    AdditionProperties additionProperties,
    SubtractionProperties subtractionProperties,
    MultiplicationProperties multiplicationProperties,
    DivisionProperties divisionProperties,
    RoundingProperties roundingProperties,
    ConversionProperties conversionProperties,
    Optional<String> subheader,
    boolean toggleHide,
    int numberOfCalculations,
    boolean verticalDisplay,
    Optional<String> exerciseReference) {
  public static final int UNLIMITED_TRANSGRESSIONS = -1;
}

File: frontend/src/shared/calculation-properties.ts
import { AdditionProperties, initialAdditionProperties } from "./addition-properties";
import { initialSubtractionProperties, SubtractionProperties } from "./subtraction-properties";
import { initialMultiplicationProperties, MultiplicationProperties } from "./multiplication-properties";
import { DivisionProperties, initialDivisionProperties } from "./division-properties";
import { initialRoundingProperties, RoundingProperties } from "./rounding-properties";
import { ConversionProperties, initialConversionProperties } from "./conversion-properties";

export interface CalculationProperties {
  operators: string[];
  additionProperties: AdditionProperties;
  subtractionProperties: SubtractionProperties;
  multiplicationProperties: MultiplicationProperties;
  divisionProperties: DivisionProperties;
  roundingProperties: RoundingProperties;
  conversionProperties: ConversionProperties;
  subheader?: string;
  toggleHide: boolean;
  numberOfCalculations: number;
  verticalDisplay: boolean;
  exerciseReference?: string;
}

export const initialCalculationProperties = {
  operators: [],
  additionProperties: initialAdditionProperties,
  subtractionProperties: initialSubtractionProperties,
  multiplicationProperties: initialMultiplicationProperties,
  divisionProperties: initialDivisionProperties,
  roundingProperties: initialRoundingProperties,
  conversionProperties: initialConversionProperties,
  toggleHide: true,
  numberOfCalculations: 15,
  verticalDisplay: false
} as CalculationProperties;
